modules/client/src/controllers/ResolveConditionController.ts

```diff
@@ -112,11 +181,89 @@ export class ResolveConditionController extends AbstractController {
     this.connext.emit("RECIEVE_TRANSFER_FINISHED_EVENT", { paymentId });
 
     return {
+      appId,
       freeBalance: await this.connext.getFreeBalance(assetId),
       paymentId,
     };
   };
 
+  private verifyInstalledApp = async (
+    appId: string,
+    amountParam: BigNumber,
+    assetIdParam: string,
+    paymentIdParam: string,
+    preImageParam: string,
+  ): Promise<void> => {
+    // get the app
+    const { appInstance } = await this.connext.getAppInstanceDetails(appId);
+
+    // verify state hash is correct
+    const {
+      linkedHash,
+      preImage,
+      coinTransfers,
+      amount,
+      assetId,
+      paymentId,
+    } = appInstance.latestState as SimpleLinkedTransferAppState;
+
+    const throwErr = (reason: string): void => {
+      throw new Error(
+        `Detected ${reason} when resolving linked transfer app ${appId}, refusing to takeAction`,
+      );
+    };
+
+    // verify initial state params are correct
+    if (!bigNumberify(amount).eq(amount)) {
+      throwErr(`incorrect amount in state`);
+    }
+
+    if (assetIdParam !== assetId) {
+      throwErr(`incorrect assetId`);
+    }
+
+    if (paymentIdParam !== paymentId) {
+      throwErr(`incorrect paymentId`);
+    }
+
+    if (linkedHash !== createLinkedHash(amountParam, assetIdParam, paymentIdParam, preImageParam)) {
+      throwErr(`incorrect linked hash`);
+    }
+
+    if (mkHash("0x0") !== preImage) {
```

Comment:
Not a huge fan of this mkHash util, the `HashZero` ethers constant would be better here IMO